R/SEIRD.R

```diff
@@ -0,0 +1,347 @@
+#'
+#' @param object an object of the class SEIRD
+#' @param times (double) a sequence of time points at which the solution to
+#' the system of ODEs should be returned. Must be of the form
+#' seq(t_start, t_end, by=t_step). Default time series is seq(0, 100, by = 1).
+#' @param solve_method (string) a string of chosen numerical integration method
+#' for solving the ode system. Default is "lsoda" which is also the default for
+#' the ode function in the deSolve package used in this function.
+#'
+#' @return a dataframe with the time steps, time series of S, E, I and R
+#' population fractions, and incidence numbers and deaths of the SEIRD model.
+#' @aliases simulate_SEIRD,ANY,ANY-method
+#' @export
+
+setMethod(
+  "simulate_SEIRD", "SEIRD",
+  function(object, times, solve_method = "lsoda") {
+    if (!is.double(times)) {
+      stop("Evaluation times of the model storage format must be a vector.")
+    }
+
+    # set initial state vector
+    state <- c(S = initial_conditions(object)$S0,
+               E = initial_conditions(object)$E0,
+               I = initial_conditions(object)$I0,
+               R = initial_conditions(object)$R0,
+               C = initial_cases_deaths(object)$C0,
+               D = initial_cases_deaths(object)$D0)
+    # set transmission parameters vector
+    parameters <- c(b = transmission_parameters(object)$b,
+                    k = transmission_parameters(object)$k,
+                    g = transmission_parameters(object)$g,
+                    m = transmission_parameters(object)$m)
+    # function for RHS of ode system
+    right_hand_side <- function(t, state, parameters) {
+      with(
+        as.list(c(state, parameters)), {
+          s <- state[1]
+          e <- state[2]
+          i <- state[3]
+          r <- state[4]
+          c <- state[5]
+          d <- state[6]
+          # rate of change
+          ds <- -b * s * i
+          de <- b * s * i - k * e
+          di <- k * e - (g + m) * i
+          dr <- g * i
+          dc <- b * s * i
+          d_death <- m * i
+          # return the rate of change
+          list(c(ds, de, di, dr, dc, d_death))
+        })
+    }
+
+    # call ode solver
+    out <- ode(
+      y = state, times = times, func = right_hand_side,
+      parms = parameters, method = solve_method)
+
+    output <- as.data.frame.array(out)
+
+    # Compute incidences and deaths
+    output$C[2:length(output$C)] <- output$C[
+      2:length(output$C)] - output$C[1:(length(output$C) - 1)]
+    output$C[1] <- 0
+    output$D[2:length(output$D)] <- output$D[
+      2:length(output$D)] - output$D[1:(length(output$D) - 1)]
+
+    colnames(output) <- c("time", object@output_names)
+
+    # Create long format of output
+    output <- melt(output, id.vars = "time")
+    names(output) <- c("time", "compartment", "value")
+
+    # Added for consistency of output format across models
+    output$age_group <- rep("0-150", length(output$time))
```

Comment:
We should ensure the long format of the output is consistent with the SEIRAge model, this looks good but we've done it slightly differently. both would probably work well.

Reply:
Thanks Oscar, I will change ours to match yours!